### PR TITLE
19 items page add sort

### PR DESCRIPTION
**Pull request by MaryAnzh**

### CODE REVIEW

**Comment by user-of-github** [comment]
Лучше, чтобы шли по очереди, а не разъезжались тут на последней строке по краям

image

File: src/core/utils/localStorage.ts [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+        } else {
+            localStorage.removeItem(key);
+        }
+    }
+
+    getData<T>(key: string): T | null {
+        const data = localStorage.getItem(key);
+
+        return data
+            ? JSON.parse(data) :
```
Comment (user-of-github) [inline src/core/utils/localStorage.ts:19]: Parsing of JSON should be wrapped with `try catch`

File: src/core/utils/localStorage.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
+export class StorageService {
+
+    constructor() { }
+
+    setData<T>(key: string, value: T | null): void {
+        if (value) {
+            localStorage.setItem(
+                key, JSON.stringify(value)
+            );
```
Comment (user-of-github) [inline src/core/utils/localStorage.ts:9]: A bit strange formatting

File: src/pages/items-page/ItemsPage.tsx [hunk 3]
```diff
@@ -1,19 +1,228 @@
...
+    brand: null | string,
+    inCart: null | string,
+}
+
+type SortType = {
+    price: null | 'assent' | 'descent',
+    store: null | 'assent' | 'descent',
+}
+
+type itemsQueryOptions = {
```
Comment (user-of-github) [inline src/pages/items-page/ItemsPage.tsx:26]: type is named from capital or small letter ?

File: src/pages/items-page/ItemsPage.tsx [hunk 4]
```diff
@@ -1,19 +1,228 @@
...
+    },
+    sort: {
+        price: null,
+        store: null,
+    },
+    search: null,
+};
+
 export const ItemsPage = (props: IItemsPageProps): JSX.Element => {
     const itemCatalog: IItem[] = [...catalog.products];
```
Comment (user-of-github) [inline src/pages/items-page/ItemsPage.tsx:46]: `ESLint: 'itemCatalog' is assigned a value but never used.(no-unused-vars)`

File: src/pages/items-page/ItemsPage.tsx [hunk 5]
```diff
@@ -1,19 +1,228 @@
...
+                    }
+                }
+            }
+        }
+        return urlParams;
+    }
+
+    //сортируем.фильтруем товары
+    modifyItemsByParams();
+    const [prods, setProds] = useState(catalogItems);
```
Comment (user-of-github) [inline src/pages/items-page/ItemsPage.tsx:180]: name `prods` doesn't tell about aim of this variable.
+ I suppose, it's better to put all `useState`s to the beginning pf component

**Comment by user-of-github** [comment]
About `ItemsPage`. It is too overloaded. So suggest at least make a separate component for Cards List.
Usually page contains just some logic and set of other components. And here page contains everything, what is not very correct and readable

**Review by user-of-github — CHANGES_REQUESTED** [review]
Needs some refactor. Especially `ItemsPage` component, which takes itself 250+ lines of code (and + types and functions before it), in future can become unreadable for maintainig

File: src/pages/items-page/ItemsPage.tsx [hunk 6]
```diff
@@ -1,19 +1,228 @@
...
+                        currentOptions.sort[subKey] = params;
+                        paramsCount++;
+                    }
+                }
+            }
+            if (objKey === 'search') {
+                const params = searchParams.get(`${objKey}`) as "assent" | "descent" | null;
+                if (params !== null) {
+                    currentOptions.search = params;
+                    paramsCount++;
```
Comment (user-of-github) [inline src/pages/items-page/ItemsPage.tsx:106]: `ESLint: 'paramsCount' is assigned a value but never used.(no-unused-vars)`

File: src/pages/items-page/ItemsPage.tsx (hunk 5 above)
Comment (MaryAnzh) [reply src/pages/items-page/ItemsPage.tsx:180]: catalogItems сортируется перед тем, как добавится.
Если есть фильтры, та там не все 100 товаров,

File: src/pages/items-page/ItemsPage.tsx [hunk 7]
```diff
@@ -1,19 +1,228 @@
...
+                        if (filtering.length > 0) {
+                            catalogItems = filtering;
+                            if (v !== '...') {
+                                urlParams.push([`${subKey}`, v]);
+                            }
+                        }
+                    }
+                }
+            }
+            if (key == 'sort') {
```
Comment (user-of-github) [inline src/pages/items-page/ItemsPage.tsx:162]: `ESLint: Expected '===' and instead saw '=='.(eqeqeq)`

File: src/pages/items-page/ItemsPage.tsx (hunk 5 above)
Comment (MaryAnzh) [reply src/pages/items-page/ItemsPage.tsx:180]: name prods doesn't tell about aim of this variable.
Имя из офиц примера по постройке магазина в реакт приложении ))))

File: src/pages/items-page/ItemsPage.tsx [hunk 8]
```diff
@@ -1,19 +1,228 @@
...
 
+type FilterType = {
+    category: null | string,
+    brand: null | string,
+    inCart: null | string,
+}
+
+type SortType = {
+    price: null | 'assent' | 'descent',
+    store: null | 'assent' | 'descent',
```
Comment (user-of-github) [inline src/pages/items-page/ItemsPage.tsx:23]: `'assent' | 'descent'` is too frequently repeated. Here and in `Sort class` as well. So maybe better put it to separate type, for example `type SortDirection`

File: src/core/utils/localStorage.ts [hunk 9]
```diff
@@ -0,0 +1,22 @@
+export class StorageService {
+
+    constructor() { }
```
Comment (user-of-github) [inline src/core/utils/localStorage.ts:3]: `ESLint: Useless constructor.(@typescript-eslint/no-useless-constructor)`

**Review by user-of-github — APPROVED** [review]

File: src/core/utils/localStorage.ts (hunk 2 above)
Comment (MaryAnzh) [reply src/core/utils/localStorage.ts:9]: Почему?
Есть и такое форматирование, так по крайней мере у тебя не сливается инструкция

Comment (user-of-github) [reply src/core/utils/localStorage.ts:9]: да, есть, но, наверное, всё же для более громоздких конструкций

**Pull request merged**